### PR TITLE
State Management (and some small updates)

### PR DESCRIPTION
**Pull request by Romero027**
1. State management (work in progress)
2. Add support for copying proto

### CODE REVIEW

**Comment by Kristoff-starling** [comment]
Just a reminder: to avoid frequent formatting commits, you could run
```
pip install pre-commit
pre-commit install
```
in the repo and it will automatically reformat the code (and ask you to re-commit) :)

File: compiler/element/backend/envoy/finalizer.py [hunk 1]
```diff
@@ -43,13 +45,12 @@ def gen_template(_placement, output_dir, snippet, lib_name, proto_path):
     # Add the proto file to the generated code
     with open(f"{template_path}/build.rs", "r") as file:
         build_file = file.read()
-    build_file = build_file.replace("PROTO_FILENAME", proto_path)
-    with open(f"{output_dir}/build.rs ", "w") as file:
+    proto = os.path.basename(proto_path).replace(".proto", "")
+    build_file = build_file.replace("PROTO_FILENAME", proto)
+    with open(f"{output_dir}/build.rs", "w") as file:
```
Comment (Kristoff-starling) [inline compiler/element/backend/envoy/finalizer.py:50]: OK, it's good to see that the whitespace has been removed (which confused me for a few hours :( )

Comment (livingshade) [reply compiler/element/backend/envoy/finalizer.py:50]: Sorry about that 😭

Comment (Kristoff-starling) [reply compiler/element/backend/envoy/finalizer.py:50]: No worries. I should have realized it earlier when VSCode refused to provide rust syntax highlight for that file 😂

**Review by livingshade — APPROVED** [review]
Looks good to me! 
It might help if we make `name2var` in `context` be a mapping from `str` to `List[Variables]`, if we have to access unwrapped state and wrapped state at the same time.  (name2var['acl'] returns both `acl_inner` and `acl` ).
I don't whether such case arises, though.

**Comment by Kristoff-starling** [comment]
@livingshade Please don't merge this draft PR for now as we plan to add external state support for the envoy backend and the 
logic of generating different codes according to consistency level hasn't been finished. :)

**Comment by Kristoff-starling** [comment]
@Romero027 I'll merge this PR first and left potential bug fixes for subsequent PRs (I'll soon submit another PR on graph compiler changes).

**Pull request merged**